### DISCUSSION TITLE
Folding and v8.4.1: how do you have this set up?

### DISCUSSION THREAD

**Discussion by fredrikaverpil** [LazyVim/LazyVim#1572] [Q&A] [answered]
I noticed folding config were added in [v8.4.0](https://github.com/LazyVim/LazyVim/releases/tag/v8.4.0) (and some tweaks in [v8.4.1](https://github.com/LazyVim/LazyVim/releases/tag/v8.4.1)). I was using [kevinhwang91/nvim-ufo](https://github.com/kevinhwang91/nvim-ufo) to enable folding prior to this release and was just wondering if maybe it is now redundant to rely on this project for treesitter-based folding...?

I really just looking to have basic folding working, using `za` to toggle folding and I'm not really looking to show arrows/indicators in the status column.  My config:

```lua
return {

  {
    "kevinhwang91/nvim-ufo",
    dependencies = {
      "kevinhwang91/promise-async",
      "nvim-treesitter/nvim-treesitter",
    },
    config = function()
      vim.o.foldcolumn = "0" -- '0' does not show the fold column, higher values increase the width
      vim.o.foldlevel = 99 -- Using ufo provider need a large value, feel free to decrease the value
      vim.o.foldlevelstart = 99
      vim.o.foldenable = true

      require("ufo").setup({
        provider_selector = function(bufnr, filetype, buftype)
          return { "treesitter", "indent" }
        end,
      })
    end,
  },
}
```

If I disable the above plugin, folding doesn't work for me when on v8.4.1. Do I have to do anything in particular to get it to work or should I stay on this config on mine?

**Comment by dpetka2001** [comment]
I also had `ufo` installed and I just disabled it and I got folds from Lazyvim without doing anything else.
![2023-10-03_18-34](https://github.com/LazyVim/LazyVim/assets/12776461/ec8fbfd5-2602-4ea5-b17e-7da01c75a76f)

**Reply by folke** [reply]
On nightly, they will look like this:

![image](https://github.com/LazyVim/LazyVim/assets/292349/f244d699-cd4e-4566-898c-ffef4629af17)

**Comment by folke** [comment]
I just pushed another fix for NEovim < 0.10, but on nigthly, your folds should look like:

![image](https://github.com/LazyVim/LazyVim/assets/292349/cf757814-98b6-4423-8b8d-4a78323824e4)

Make sure to delete any fold options like `foldtext` and `statuscolumn` that you may have set in your own config.

**Comment by fredrikaverpil** [comment]
I'm not sure why, but with ufo disabled and uninstalled; if I launch `nvim` and then open up a file with `<leader>space` (`<leader><leader>`), I can't fold anything.

But if I open up a second file and switch back to the first one, then I can fold in the first file with `za`.

Not sure why the behavior is like that here. I don't have any `foldtext` or `statuscolumn` in my own config. Do you experience this behavior too?

**Reply by folke** [reply]
I actually noticed that a couple of times as well, but it's hard to reproduce.
I think it's caused by the foldexpr running before a document's parser is loaded. Or before nvim-treesitter was loaded.
Will see if I can fix this.

**Reply by folke** [reply]
Have you tried with the latest version from a couple of minutes ago? I don't seem to be able to reproduce this

**Reply by fredrikaverpil** [reply]
Yes, I'm on version 8.4.3 now and I still see this weird behavior where I get a "E490: No fold found" notification when opening the first file and hitting `za`. Switching back and forth to another file once fixes it.

**Reply by fredrikaverpil** [reply]
To be precise:

1. Open file 1: cannot fold inside it
2. Open file 2: cannot fold inside it
3. Switch back to file 1: can successfully fold inside it
4. Switch to file 2: can successfully fold inside it

**Reply by folke** [reply]
yep, can reproduce. will check

In the first file, if you then do `:set foldmethod=expr`, then it all works again

**Reply by fredrikaverpil** [reply]
I don't know if it helps but this is my treesitter config, in case that has something to do with this: https://github.com/fredrikaverpil/dotfiles/blob/main/nvim-lazyvim/lua/plugins/treesitter.lua

**Reply by fredrikaverpil** [reply]
> In the first file, if you then do :set foldmethod=expr, then it all works again

yes, can repro that here too 😄

**Reply by folke** [reply]
It only happens with files opened through telescope, not for any other way of opening a file

**Reply by folke** [reply]
https://github.com/nvim-telescope/telescope.nvim/issues/699

**Reply by folke** [reply, marked as answer]
Should be fixed now: https://github.com/LazyVim/LazyVim/commit/ccff8683ba957250d71ef66429fc877f0ee5cedb

Will make a new release in a bit

**Comment by skoch13** [comment]
On a side note, how do I configure automatic folding for imports as nvim-ufo provides?

**Reply by folke** [reply]
no idea. the LazyVim changes are using Neovim treesitter folds directly.
You can probably write a function to do what you want, but I'm not planning to add this functionality to LazyVim.

**Reply by dimas-cyriaco** [reply]
@skoch13 this works for me:

```lua
return {
  "kevinhwang91/nvim-ufo",
  dependencies = { "kevinhwang91/promise-async" },
  config = {
    close_fold_kinds = { "imports" },
  },
}
```

**Reply by skoch13** [reply]
Thanks, that's how it configured at the moment. My question was about the treesitter foldings, not ufo's 😀

**Comment by fredrikaverpil** [comment]
Is it just me or is folding ~~completely~~ broken now?
I'm on LazyVim [v10.5.0](https://github.com/LazyVim/LazyVim/releases/tag/v10.5.0).

EDIT: I've tried opening both Python files and Go files and I can't fold using `<leader>za` in those. It seems I can fold in Lua files though.

**Reply by fredrikaverpil** [reply]
If [enabling ufo](https://github.com/fredrikaverpil/dotfiles/blob/1222633551e143157638ba791287d3d247d65562/nvim-lazyvim/lua/plugins/fold.lua#L5), folding works fine in Python and Go files, so I guess I'll do that.

**Reply by dpetka2001** [reply]
Probably has to do with your Neovim version. I'm on nightly and I don't observe such behavior. Ufo tries to create folds using LSP as first pattern if I remember correctly. For Neovim 0.9 `foldmethod` is set to `indent` instead of `expr` and doesn't use treesitter, so that may be the reason for the behavior you're observing.

**Reply by fredrikaverpil** [reply]
> Probably has to do with your Neovim version.

Yeah that is maybe it. I'm on v0.9.4.

**Reply by fredrikaverpil** [reply]
I'm just pretty sure it used to work when Folke was done with the addition of it... but 🤷 it works for me now that I have ufo enabled.

**Reply by dpetka2001** [reply]
No, he disabled treesitter expr in folding, because it was causing a lot of crushes to users on stable. See [here](https://github.com/LazyVim/LazyVim/commit/d989ecc943b3240db0be3aa0369b96089cee4b40). I suggest you switch to `bob` to install Neovim and be able to freely change between stable and nightly.

**Comment by kobzar** [comment]
I can advice you this plugin
https://github.com/chrisgrieser/nvim-origami
This is best all of i tryed for all time.
